### ISSUE TITLE
[Source Generators] Make it easier to filter down to the syntax nodes you care about

### ISSUE THREAD

**Issue by tannergooding** [dotnet/roslyn#47842]
# Issue

It is currently non-trivial to filter which syntax nodes you want to process.

# Rationale

Currently if you want to generate code for a partial method, say:
```csharp
[SomeAttribute]
public partial void MyMethod();
```

Your generator is going to want to only process partial `MethodDeclarationSyntax` nodes that have `SomeAttribute`.

Today you can use `RegisterForSyntaxNotifications` to register a syntax receiver which gets invoked for every `SyntaxNode` in the compilation. However, you only have the `SyntaxNode` and while you can trivially do:
```csharp
if (syntaxNode is not MethodDeclarationSyntax methodDeclarationSyntax)
{
    return;
}

if (!methodDeclarationSyntax.Modifiers.Any(SyntaxKind.PartialKeyword))
{
    return;
}

if (methodDeclarationSyntax.AttributeLists.Count == 0)
{
    return;
}

// rest of the logic
```

It is then non-trivial to confirm that one of the attributes in one of the attribute lists is the attribute you care about. There may be using aliases or fully qualified names involved which mean that a simple check for `attribute.Name == "SomeAttribute"` is not sufficient and you must consider/handle various edge cases to get it correct.

Alternatively, you could just take all partial methods that have an attribute, but this could be potentially many "false positives".
You might consider filtering based on the number of arguments an attribute has, but given properties and optional parameters (such as exist for `DllImport`) this is again potentially many false positives.

This also means that you incur traversal cost for nodes that may have zero relevance to you, such as `enums` or `delegates` which can't contain partial methods or you might traversal all `methods` when you only care about `fields`, etc.

# Proposal

A set of APIs should be provided that make it simpler to filter down to just the syntax nodes you care about, reducing false positives and overall allocations required by a given source generator.

It might likewise be beneficial to provide a way to indicate it is not necessary to traverse a given syntax node. For example, if you only care about methods, and a given type contains no methods or members that could contain methods, then it is pointless to visit the remaining child nodes for the given generator.

**Comment by tannergooding** [comment]
A case where this is inefficient is for `TerraFX.Interop.Windows`. Doing some basic analysis reports: `MethodDeclarations: 25,713; NonMemberDeclarations: 2,595,195; Nodes: 2,703,020`
* `MethodDeclarations` is any node where `syntaxNode is MethodDeclarationSyntax`
* `NonMemberDeclarations` is any node where `syntaxNode is not MemberDeclarationSyntax`
* `Nodes` is just all nodes traversed

and while I'm not sure if there are any edge cases where a `MethodDeclarationSyntax` can be the child of something that isn't a `MemberDeclarationSyntax`; I think it does show the potentiality for traversing nodes unnecessarily.

**Comment by jinujoseph** [comment]
cc @jasonmalinowski , @chsienki

**Comment by jasonmalinowski** [comment]
@chsienki Is there any reason not to have (at least an option) to get a semantic model along with the node you're registering for when registering for syntax?

**Comment by chsienki** [comment]
@jasonmalinowski at the point of the SyntaxWalk we haven't done any binding yet, so there is no sematic model to get. The idea is that you can collect the 'interesting' things you want to ask questions about during a shared walk (all generators share the same walk), then ask the semantic model questions about it during execution. 

Ideally we'd use the analyzer model of walking through the bound nodes, but that requires us to realize the whole compilation twice, every time, even if none of the generators end up using those nodes.

**Comment by alrz** [comment]
>Ideally we'd use the analyzer model of walking through the bound nodes, but that requires us to realize the whole compilation twice

If I understand this right, does that mean analyzers and generators should have a common interface to do this in one pass?

**Comment by jasonmalinowski** [comment]
@alrz: the generator and analyzer passes (even if they're a common API) would still be separate passes, since we always run generators first, produce the new Compilation, and then run analyzers. Otherwise the analyzers are running on incomplete code.

@chsienki: that intent makes sense, I'm trying to figure out what the generator is expected to do though. In @tannergooding's case, all they could do is collect all methods with attributes per their snippet, and then in their execute go and bind every method, without any semantic model caching to benefit them. Does there need to be a syntax-only pass and then they can subscribe for a semantic pass that does a second walk or something?

**Comment by chsienki** [comment]
@jasonmalinowski  Basically what you describe is the intent. Grab everything, then go ask questions. 

_Today_ We don't have any caching, but we do have a plan to fix that https://github.com/dotnet/roslyn/projects/54#card-37390612 the fix itself should be transparent to the user, so we're recommending they just use the semantic model as-is today.

The advantage there is we'll only realize the minimum required per-generator. Doing a walk would still require double binding everything.

**Comment by CyrusNajmabadi** [comment]
> Your generator is going to want to only process partial MethodDeclarationSyntax nodes that have SomeAttribute.

This is now handled by 'ForAttributeWithMetadataName'.
